src/pathfinding_service/api/rest.py

```diff
@@ -120,6 +147,102 @@ def post(self, token_network_address: str):
+        raise exceptions.WrongIOURecipient(expected=pathfinding_service.address)
+    if not iou.is_signature_valid():
+        raise exceptions.InvalidSignature
+
+    # Compare with known IOU
+    active_iou = pathfinding_service.database.get_iou(
+        sender=iou.sender,
+        claimed=False,
+    )
+    if active_iou:
+        if active_iou.expiration_block != iou.expiration_block:
+            raise exceptions.UseThisIOU(iou=active_iou)
+
+        expected_amount = active_iou.amount + pathfinding_service.service_fee
+    else:
+        if pathfinding_service.database.get_iou(
+            sender=iou.sender,
+            expiration_block=iou.expiration_block,
+            claimed=True,
+        ):
+            raise exceptions.IOUAlreadyClaimed
+
+        min_expiry = pathfinding_service.web3.eth.blockNumber + MIN_IOU_EXPIRY
+        if iou.expiration_block < min_expiry:
+            raise exceptions.IOUExpiredTooEarly(min_expiry=min_expiry)
+        expected_amount = pathfinding_service.service_fee
+    if iou.amount < expected_amount:
+        raise exceptions.InsufficientServicePayment(expected_amount=expected_amount)
+
+    # TODO: deposit large enough?
+
+    # Save latest IOU
+    iou.claimed = False
+    pathfinding_service.database.upsert_iou(iou)
+
+
+@add_schema
+@dataclass
+class IOURequest:
+    """A HTTP request to IOUResource"""
+    sender: Address
+    receiver: Address
+    timestamp: datetime
+    timestamp_str: str = field(metadata={
+        "marshmallow_field": marshmallow.fields.String(load_from='timestamp'),
+    })
+    signature: Signature = field(metadata={"marshmallow_field": HexedBytes()})
+    Schema: ClassVar[Type[marshmallow.Schema]]
+
+    def is_signature_valid(self):
+        packed_data = (
+            Web3.toBytes(hexstr=self.sender) +
+            Web3.toBytes(hexstr=self.receiver) +
+            Web3.toBytes(text=self.timestamp_str)
+        )
+        try:
+            recovered_address = eth_recover(packed_data, self.signature)
+        except InvalidSignature:
+            return False
+        return to_checksum_address(recovered_address) == to_checksum_address(self.sender)
+
+
+class IOUResource(PathfinderResource):
+
+    def get(self, token_network_address: TokenNetworkAddress):
+        iou_request, errors = IOURequest.Schema().load(request.args)
+        if errors:
+            raise exceptions.InvalidRequest(**errors)
+        if not iou_request.is_signature_valid():
+            raise exceptions.InvalidSignature
+        if iou_request.timestamp < datetime.utcnow() - timedelta(hours=1):
+            raise exceptions.RequestOutdated
+
+        last_iou = self.pathfinding_service.database.get_iou(
+            sender=iou_request.sender,
+            claimed=False,
+        )
+        if last_iou:
+            last_iou = IOU.Schema(strict=True, exclude=['claimed']).dump(last_iou)[0]
+
+        return {
+            'last_iou': last_iou,
+        }, 200
```

Comment:
I wonder if we should return a 404 Not Found here, if no IOU is available.

Reply:
Yes, I that's an even better fit than 200.